src/Http/Controllers/Subscribers/SubscribersImportController.php

```diff
@@ -48,30 +50,103 @@ public function store(SubscribersImportRequest $request): RedirectResponse
         if ($request->file('file')->isValid()) {
             $filename = Str::random(16) . '.csv';
 
-            $path = $request->file('file')->storeAs('imports', $filename);
+            $path = $request->file('file')->storeAs('imports', $filename, 'local');
 
-            $subscribers = (new FastExcel)->import(storage_path('app/' . $path), function (array $line) use ($request) {
-                // TODO: validate each row beforehand
+            $errors = $this->validateCsvContents(Storage::disk('local')->path($path));
+
+            if (count($errors->getBags())) {
+                Storage::disk('local')->delete($path);
+
+                return redirect()->back()
+                    ->withInput()
+                    ->with('error', __('The provided file contains errors'))
+                    ->with('errors', $errors);
+            }
+
+            $errors = new ViewErrorBag();
+
+            $row = 1;
+
+            $subscribers = (new FastExcel)->import(Storage::disk('local')->path($path), function (array $line) use ($request, $errors, &$row) {
                 try {
                     $data = Arr::only($line, ['id', 'email', 'first_name', 'last_name']);
 
+                    $this->validateData($data);
+
                     $data['segments'] = $request->get('segments') ?? [];
 
+                    $row++;
+
                     return $this->subscriberService->import(auth()->user()->currentWorkspace()->id, $data);
-                } catch (Exception $e) {
-                    Log::warning($e->getMessage());
+                } catch (ValidationException $e) {
+                    $errors->put('Row ' . $row, $e->validator->errors());
+
+                    $row++;
                 }
 
                 return null;
             });
 
-            Storage::disk('local')->delete('imports/' . $filename);
+            Storage::disk('local')->delete($path);
 
-            return redirect()->route('sendportal.subscribers.index')
-                ->with('success', __('Imported :count subscriber(s)', ['count' => $subscribers->count()]));
+            if (empty($errors->getBags())) {
+                return redirect()->route('sendportal.subscribers.index')
+                    ->with('success', __('Imported :count subscriber(s)', ['count' => $subscribers->count()]));
```

Comment:
This could be confusing if you are only updating subscribers. How difficult would it be to change it to something like;
"Success! Imported **3** subscribers and updated **5** subscribers."

Reply:
I think this is also possible.